File: Sample/Sample/Settings View/TagsTableViewController.m
/* Copyright Airship and Contributors */

@import AirshipCore;
@import AirshipLocation;

#import "TagsTableViewController.h"

@implementation TagsTableViewController

NSString *addTagsSegue = @"addTagsSegue";

- (void)viewDidLoad {
    [super viewDidLoad];
    
    UIBarButtonItem* addButton = [[UIBarButtonItem alloc] initWithBarButtonSystemItem:UIBarButtonSystemItemAdd target:self action:@selector(addTag)];
    
    self.navigationItem.rightBarButtonItem = addButton;
}

- (void)setTableViewTheme{
    self.tableView.backgroundColor = [UIColor whiteColor];
    [self.navigationController.navigationBar setTitleTextAttributes: @{NSForegroundColorAttributeName:[UIColor whiteColor]}];
    self.navigationController.navigationBar.barTintColor = [UIColor blackColor];
}

- (void)viewWillAppear:(BOOL)animated {
    [super viewWillAppear:(animated)];
    [self.tableView reloadData];
}

- (void)addTag {
    [self performSegueWithIdentifier:addTagsSegue sender:self];
}

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
    return 1;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return UAirship.channel.tags.count;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    static NSString *cellIdentifier = @"tagCell";
    
    UITableViewCell *cell = [self.tableView dequeueReusableCellWithIdentifier:cellIdentifier];
    
    if (cell == nil) {
        cell = [[UITableViewCell alloc] initWithStyle:UITableViewCellStyleDefault reuseIdentifier:@"tagCell"];
    }
    cell.textLabel.text = UAirship.channel.tags[indexPath.row];
    cell.textLabel.textColor = [UIColor blackColor];
    cell.detailTextLabel.textColor = [UIColor colorWithRed:0.51 green:0.51 blue:0.53 alpha:1.0];
    cell.backgroundColor = [UIColor whiteColor];
    
    return cell;
}

- (void)tableView:(UITableView *)tableView commitEditingStyle:(UITableViewCellEditingStyle)editingStyle forRowAtIndexPath:(NSIndexPath *)indexPath {
    if (editingStyle == UITableViewCellEditingStyleDelete && [tableView cellForRowAtIndexPath:indexPath].textLabel.text != nil) {
        UATagEditor *tagEditor = [UAirship.channel editTags];
        [tagEditor removeTag:[tableView cellForRowAtIndexPath:indexPath].textLabel.text];
        [tagEditor apply];
        [tableView deleteRowsAtIndexPaths:@[indexPath] withRowAnimation: UITableViewRowAnimationFade];
    }
}

@end

File: SwiftSample/SwiftSample/TagsTableViewController.swift
/* Copyright Airship and Contributors */

import UIKit

#if canImport(AirshipCore)
import AirshipCore
#elseif canImport(Airship)
import Airship
#endif

class TagsTableViewController: UITableViewController {
    let addTagsSegue:String = "addTagsSegue"

    override func viewDidLoad() {
        super.viewDidLoad()

        let addButton:UIBarButtonItem = UIBarButtonItem(barButtonSystemItem: UIBarButtonItem.SystemItem.add, target: self, action: #selector(TagsTableViewController.addTag))
        navigationItem.rightBarButtonItem = addButton
    }

    func setTableViewTheme() {
        tableView.backgroundColor = #colorLiteral(red: 0.999904573, green: 1, blue: 0.9998808503, alpha: 1)
        navigationController?.navigationBar.titleTextAttributes = [.foregroundColor:#colorLiteral(red: 0.999904573, green: 1, blue: 0.9998808503, alpha: 1)]
        navigationController?.navigationBar.barTintColor = #colorLiteral(red: 0, green: 0, blue: 0, alpha: 1);
    }

    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated);
        setTableViewTheme()
        tableView.reloadData()
    }

    @objc func addTag () {
        performSegue(withIdentifier: addTagsSegue, sender: self)
    }

    override func numberOfSections(in tableView: UITableView) -> Int {
        return 1
    }

    override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return Airship.channel.tags.count
    }

    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        var cell = tableView.dequeueReusableCell(withIdentifier: "tagCell", for: indexPath)

        if cell.isEqual(nil) {
            cell = UITableViewCell.init(style: UITableViewCell.CellStyle.default, reuseIdentifier:"tagCell")
        }
        cell.textLabel!.text = Airship.channel.tags[indexPath.row]
        cell.textLabel?.textColor = #colorLiteral(red: 0, green: 0, blue: 0, alpha: 1)
        cell.detailTextLabel?.textColor = #colorLiteral(red: 0.513617754, green: 0.5134617686, blue: 0.529979229, alpha: 1)
        cell.backgroundColor = #colorLiteral(red: 0.999904573, green: 1, blue: 0.9998808503, alpha: 1)

        return cell
    }

    override func tableView(_ tableView: UITableView, commit editingStyle: UITableViewCell.EditingStyle, forRowAt indexPath: IndexPath) {
        if (editingStyle == .delete &&
            tableView.cellForRow(at: indexPath)?.textLabel?.text?.isEmpty == false) {

            Airship.channel.editTags { editor in
                editor.remove((tableView.cellForRow(at: indexPath)?.textLabel?.text)!)
            }
            tableView.deleteRows(at: [indexPath], with: .fade)

            Airship.push.updateRegistration()
        }
    }
}
